enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,616 @@
+  so that I can get bug fixes, new features, and security patches.
+* As a MicroShift administrator, I want automated backups of 
+  MicroShift data so in case of a greenboot failure system will rollback
+  and backup will be restored.
+
+### Goals
+
+Goal of the enhancement is to describe implementation roadmap for
+integrating MicroShift with ostree and greenboot in order to provide
+functionality to:
+- Safely update MicroShift version (by backing up the data and
+  restoring it in case of rollback)
+- Migrating internal data (like Kubernetes storage or etcd schema) to
+  newer version
+- Block upgrades in case of version bump being too big
+
+Design aims to implement following principles:
+- Keep it simple, optimize later
+- MicroShift does not own the OS or host
+- MicroShift and all its components are version, upgraded, and rolled back together
+- Be defensive, fail fast
+- Rely on outside intervention as a last resort
+
+### Non-Goals
+
+* Building allowed/blocked version migration graph
+* Handling readiness, and backup and rollback of 3rd party applications
+  (although end user documentation should be provided)
+* Defining updateability for non-ostree systems is left to a future enhancement
+* Protecting against data corruption
+
+## Proposal
+
+### Workflow Description
+
+#### User perspective
+
+**MicroShift administrator** is a human responsible for preparing
+ostree commits and scheduling devices to use these commits.
+
+1. MicroShift administrator prepares a new ostree commit
+1. MicroShift administrator schedules device to reboot and use new ostree commit
+1. Device boots to new commit
+1. Operating System, greenboot, and MicroShift take actions without any additional intervention
+
+#### Implementation perspective
+
+**First boot into commit with MicroShift**
+
+1. First ostree commit is installed
+1. First ostree commit boots
+1. MicroShift starts
+   - No backup action (did not previously run, so there is nothing to backup, and did not previously fail so no restore needed)
+1. MicroShift startup succeeds
+1. All greenboot checks pass
+1. Greenboot runs green scripts
+   - MicroShift green script sets backup mode to "backup"
+
+**Simple host reboot (cont of previous flow)**
+
+1. First ostree commit shuts down
+1. First ostree commit boots
+1. MicroShift starts up
+   - Backup mode is "backup",
+   - Backup script runs, creating a backup compatible with the first ostree commit
+1. MicroShift startup succeeds
+1. All greenboot checks pass
+1. Greenboot runs green scripts
+   - MicroShift green script sets backup mode to "backup" (no change)
+
+**Failed upgrade (cont of previous flow)**
+
+1. Second ostree commit is staged
+1. First ostree commit shuts down
+1. Second ostree commit boots
+1. MicroShift starts up
+   - Backup mode is "backup",
+   - Backup script runs, creating a backup compatible with the first ostree commit
+1. MicroShift startup fails
+1. Greenboot runs red scripts
+    - Set backup mode to "restore"
```

Comment:
We should only set restore on the last greenboot attempt.
See [this line](https://github.com/openshift/microshift/blob/a45ce756d09e95e9102bbf016bbd201cd37d8e0a/packaging/greenboot/microshift-pre-rollback.sh#L38) on how MicroShift greenboot identifies that case.

Reply:
But this is based on assumption from https://github.com/openshift/enhancements/pull/1387#discussion_r1174612531, right?
If subsequent boots fail, but they implicate "restore", then we're technically correct.
Would you consider this an optimization?

Reply:
Also, I'd like to avoid depending on grub vars, they seem very ephemeral.
`greenboot-grub2-set-success.service` unsets the `boot_counter` and I'd like to avoid complicated logic handle such "race conditions"

Reply:
Understood and agreed that we should avoid external dependencies as much as possible.

Reply:
If components other than MicroShift fail (and unrelated to it), red scripts will still run but we should not change to restore, right? In the end MicroShift is another component, but if everything looks ok then you should not rollback.